release-process.md

```diff
@@ -1,54 +1,131 @@
+  - Code starts out in a pull request which is reviewed by other team members (see [Code Review](code-review.md)).
+  - Then the pull request is merged to the `main` branch (or `master` for legacy projects).
+
+### Process
+
+```mermaid
+sequenceDiagram
+  actor eng as Software Engineers
+  participant branchmain as main/master
+
+  participant branchrc as release-candidate
+  participant envrc as Release Candidate
+
+  participant branchrelease as release
+  participant envprod as Production
+
+  loop
+    eng->>branchmain: Merge pull requests
+  end
+
+  Note over eng: Requests a release candidate
+
+  critical New Release Candidate
+    branchmain->>branchrc: Latest code merged
+    branchrc->>branchrc: Release notes and PR generated
+    branchrc->>envrc: Deploy to RC
+    eng->>envrc: Engineer verifies their commit(s)
+    note over eng: Checks off boxes on the PR
+  end
+
+  critical Production Release
+    eng->>branchrc: Engineer coordinated release approval
+    branchrc->>branchmain: Merged back to main
+    branchmain->>branchrelease: Main branch merged to release
+    branchrelease->>envprod: Deploy to Production
+  end
+
+
+```
+
+#### CI
+
+  - Deployments to the CI environment are only applicable to web applications.
+  - After tests pass on the `master`/`main` branch this triggers a deployment to a CI (continuous integration) server for web application projects.
+
+#### Release candidate
+  
+  - On a daily basis each project is checked on an automated schedule for new changes:
+    - If there are changes a release candidate is created in the `release-candidate` branch.
+    - A new commit with updated release notes and a version bump is made on that branch.
+    - A PR is created on GitHub to merge these changes into `release`.
+    - **Web Applications only:** Once tests pass a deployment to the RC (release candidate) server for web application projects.
+  - Unit tests are run against the release candidate and team members check off checkboxes to verify,
+    at a glance, that their code works in conjuction with other code in the release.
+  - The release candidate is merged into the `release` branch. Heroku and possibly other build systems
+    will detect this and trigger a deployment to a production server for web application projects.
+  - Library packages may need to be published to a repository like PyPI or NPM. This happens automatically once tests pass.
+## Versioning
+
+For **library projects**, our releases are identified by a version number that follows the
 [Semantic Versioning scheme]([http://semver.org/).
 
-## Release Automation Script
-The release process is partially automated through a BASH script.  The script
-is available in a public GitHub repository: [mitodl/release-script](https://github.com/mitodl/release-script)
+For **application projects**, our releases are roughly versioned by incrementing the minor version every week (resetting patch to `0`) and the patch version every release during that week.
```

Comment:
I'm definitely for that, it would make it a lot easier to reason about those numbers, not to mentioned that the convention here is a bit wonky cause it's more like "on Mondays we do a minor bump, on other days we do a patch bump", which means if the set of changes for a project isn't on a Monday, it doesn't get the minor bump.

Reply:
I've always thought this versioning method was strange. The minor / patch version numbers are supposed to be used to gate breaking changes. In Node for example, if I have an application that depends on a library, I can usually feel confident that I can do something like `~1.0.4` and any installs will grab the latest patch version, but go no higher than that. Since we update the minor versions of our packages every week, it fundamentally breaks that assumption. Once next week rolled around and 1.1.0 comes out, that app would no longer receive updated versions of that package, whether it contained breaking changes or not. Similarly, if you used `^1.0.4`, your app would start installing every minor version update. Since we are updating this every week, the only way to gate breaking changes is behind a major version number, something I'm not even sure Doof can do unless you start a release with a manual version number. As @blarghmatey says, the version number doesn't even convey any helpful information the way we're using it. If you're reading a manufacturer date code on a piece of tech for example, at least the number there can tell you exactly when the gear was manufactured. Ours tells you potentially the amount of weeks since the first release, which is not a readily available piece of information. It might not even be totally accurate still because there are weeks that we don't work, like around Christmas. I'm not sure what the right answer is here, but I think more thought should be put into version numbers. Just spitballing here, but it would have to be a developer's responsibility somehow to indicate in their PR that it contains a breaking change. We could potentially do that with a checkbox or something, and if a breaking change is detected in the release then the minor version is automatically updated.

Reply:
Date based versioning for apps and semver for libraries seems to make sense to me.

Reply:
It makes sense to me as well. It's how we did in my last company, like this format: 2022-09-28.2 in case there are 2 releases for the day. It's also good for reporting and metrics if we want to know how many releases we have done during some period of time